crates/driver/src/domain/competition/auction.rs

```diff
@@ -55,31 +92,111 @@ impl Auction {
+        let mut balances: HashMap<
+            (order::Trader, eth::TokenAddress),
+            Result<eth::TokenAmount, crate::infra::blockchain::Error>,
+        > = join_all(tokens_by_trader.map(|(trader, token)| async move {
+            let balance = eth.balance_of(trader.into(), token).await;
+            ((trader, token), balance)
+        }))
+        .await
+        .into_iter()
+        .collect();
+
+        self.orders.retain(|order| {
+            // TODO: We should use balance fetching that takes interactions into account
+            // from `crates/shared/src/account_balances/simulation.rs` instead of hardcoding
+            // an Ethflow exception. https://github.com/cowprotocol/services/issues/1595
+            if Some(order.signature.signer.0) == eth.contracts().ethflow_address().map(|a| a.0) {
+                return true;
+            }
+
+            let remaining_balance = match balances
+                .get_mut(&(order.trader(), order.sell.token))
+                .unwrap()
+            {
+                Ok(balance) => &mut balance.0,
+                Err(err) => {
+                    let reason = observe::OrderExcludedFromAuctionReason::CouldNotFetchBalance(err);
+                    observe::order_excluded_from_auction(order, reason);
+                    return false;
+                }
+            };
 
-        Ok(Self {
-            id,
-            orders,
-            tokens,
-            gas_price,
-            deadline,
-        })
-    }
+            fn max_fill(order: &competition::Order) -> anyhow::Result<U256> {
+                use {
+                    anyhow::Context,
+                    shared::remaining_amounts::{Order as RemainingOrder, Remaining},
+                };
+
+                let remaining = Remaining::from_order(&RemainingOrder {
+                    kind: match order.side {
+                        order::Side::Buy => model::order::OrderKind::Buy,
+                        order::Side::Sell => model::order::OrderKind::Sell,
+                    },
+                    buy_amount: order.buy.amount.0,
+                    sell_amount: order.sell.amount.0,
+                    fee_amount: order.fee.user.0,
+                    executed_amount: match order.partial {
+                        order::Partial::Yes { executed } => executed.0,
+                        order::Partial::No => 0.into(),
+                    },
+                    partially_fillable: match order.partial {
+                        order::Partial::Yes { .. } => true,
+                        order::Partial::No => false,
+                    },
+                })
+                .context("Remaining::from_order")?;
+                let sell = remaining
+                    .remaining(order.sell.amount.0)
+                    .context("remaining_sell")?;
+                let fee = remaining
+                    .remaining(order.fee.user.0)
+                    .context("remaining_fee")?;
```

Comment:
Honestly it's not really clear to me what `remaining.remaining` is supposed to do, but I guess one is expected to dig into the code to understand this?

Reply:
Ideally you shouldn't have to dig into the code. But the naming for this type and function is not straight forward.
Basically the idea is that the `remaining` variable holds all the information needed to know how much of an order was executed. And the `remaining` method then allows you to scale the initial amount to the amount that is currently left in the order.